src/main/java/com/microsoft/sqlserver/jdbc/SQLQueryMetadataCache.java

```diff
@@ -0,0 +1,272 @@
+            } catch (Exception e) {
+                throw new SQLServerException(null, SQLServerException.getErrString
+                    ("R_CEKDecryptionFailed"), null, 0, false);
+            }
+        }
+        
+        Map<Integer, CekTableEntry> enclaveKeys = session.getCryptoCache().getEnclaveEntry(encryptionValues.getValue());
+        return (enclaveKeys == null);
+    }
+
+    
+    /**
+    * 
+    * Adds the parameter metadata to the cache, also handles cache trimming.
+    * 
+    * @param params
+    *        List of parameters used
+    * @param parameterNames
+    *        Names of parameters used
+    * @param session
+    *        Enclave session containing the cryptocache
+    * @param connection
+    *        SQLServerConnection
+    * @param stmt
+    *        SQLServer statement used to retrieve keys to find correct cache
+    * @param cekList
+    *        The list of CEKs (from the first RS) that is also added to the cache as well as parameter metadata
+    * @return true, if the query metadata has been added correctly
+    */
+    static boolean addQueryMetadata(Parameter[] params, ArrayList<String> parameterNames, EnclaveSession session,
+            SQLServerConnection connection, SQLServerStatement stmt, Map<Integer, 
+            CekTableEntry> cekList) throws SQLServerException {
+
+        AbstractMap.SimpleEntry<String, String> encryptionValues = getCacheLookupKeys(stmt, connection);
+        if (encryptionValues.getKey() == null) {
+            return false;
+        }
+
+        ConcurrentHashMap<String, CryptoMetadata> metadataMap = new ConcurrentHashMap<>(params.length);
+        
+        for (int i = 0; i < params.length; i++) {
+            try {
+                CryptoMetadata cryptoCopy = null;
+                CryptoMetadata metaData = params[i].getCryptoMetadata();
+                if (metaData != null) {
+
+                    cryptoCopy = new CryptoMetadata(metaData.getCekTableEntry(), metaData.getOrdinal(),
+                            metaData.getEncryptionAlgorithmId(), metaData.getEncryptionAlgorithmName(),
+                            metaData.getEncryptionType().getValue(), metaData.getNormalizationRuleVersion());
+                }
+                if (cryptoCopy != null && !cryptoCopy.isAlgorithmInitialized()) {
+                    String paramName = parameterNames.get(i);
+                    metadataMap.put(paramName, cryptoCopy);
+                } else {
+                    return false;
+                }
+            } catch (SQLServerException e) {
+                throw new SQLServerException(null, SQLServerException.getErrString
+                                                    ("R_cryptoCacheInaccessible"), null, 0, false);
+            }
+        }
+        
+        // If the size of the cache exceeds the threshold, set that we are in trimming and trim the cache accordingly.
+        int cacheSizeCurrent = session.getCryptoCache().getParamMap().size();

```

Comment:
can we save the cryptoCache instead so we don't have to call getCryptoCache() so many times?